Added Cloudwatt driver

ZuluPro opened the pull request:
Added Driver for new French cloud provider Cloudwatt https://www.cloudwatt.com/fr/

This provider is based on OpenStack but uses a custom kind of authentication,
The request data must look like below:
{'auth': {
  'passwordCredentials': {
    'username': 'THE_USERNAME',
    'password': 'THE_PASSWORD'
  },
  'tenantId': 'THE_TENANT_ID
}}

So, I adapt classes to accept tenant_id argument.

I manually tested it with a custom script and `Driver.list_nodes()`.

ZuluPro committed:
Added Cloudwatt driver

ZuluPro committed:
Flaked8

sebgoa commented on libcloud/compute/types.py:
```diff
@@ -121,6 +121,7 @@ class Provider(object):
     IKOULA = 'ikoula'
     OUTSCALE_SAS = 'outscale_sas'
     OUTSCALE_INC = 'outscale_inc'
+    CLOUDWATT = 'cloudwatt'
```
Hi, can you put this under the #OpenStack based providers ….couple lines below ?

sebgoa commented on libcloud/compute/drivers/cloudwatt.py:
```diff
@@ -0,0 +1,149 @@
+        :param tenant_id: ID of tenant required for Cloudwatt auth
+        :type tenant_id: ``str``
+
+        Note: tenant_name argument is required for HP cloud.
```
Maybe replace HP cloud with Cloudwatt cloud ?

sebgoa commented on libcloud/compute/drivers/cloudwatt.py:
```diff
@@ -0,0 +1,149 @@
+    def __init__(self, key, secret, tenant_id, secure=True, tenant_name=None,
+                 host=None, port=None, **kwargs):
+        """
+        @inherits:  :class:`NodeDriver.__init__`
```
inherits CloudwattNodeDriver ?

sebgoa replied:
Never mind, my bad

ZuluPro committed:
Sort Cloudwatt type & fix pydoc

Kami commented on libcloud/compute/drivers/cloudwatt.py:
```diff
@@ -0,0 +1,147 @@
+# contributor license agreements.  See the NOTICE file distributed with
```
Missing first line of the license header - `# Licensed to the Apache Software Foundation (ASF) under one or more`

Kami commented on libcloud/compute/drivers/cloudwatt.py:
```diff
@@ -0,0 +1,147 @@
+Cloudwatt driver.
+"""
+import sys
+import json
```
Please use

``` python
try:
    import simplejson as json
except ImportError:
    import json
```

Otherwise it won't work with Python 2.5.

Kami commented on libcloud/compute/drivers/cloudwatt.py:
```diff
@@ -0,0 +1,147 @@
+    name = 'Cloudwatt Auth'
+
+    def __init__(self, *args, **kwargs):
+        self._ex_tenant_id = kwargs.pop('ex_tenant_id')
```
Hm, could we maybe use an existing `ex_tenant_name` argument here?

Kami commented on libcloud/compute/drivers/cloudwatt.py:
```diff
@@ -0,0 +1,147 @@
+    website = 'https://www.cloudwatt.com/'
+    connectionCls = CloudwattConnection
+    auth_url = BASE_URL
+    service_type = 'compute'
```
IIRC, `service_type` and `auth_url` should be set on the Connection and not NodeDriver class.

Kami commented:
@ZuluPro Let me know once those comments have been addressed and I will get the PR merged.

ZuluPro commented:
Ok, thx! I'm not dead only in holydays...

Kami commented:
@ZuluPro Hey, any progress on this?

It would be great if those comments would get addressed so the driver can be included in the next release.

ZuluPro committed:
Fix license 1st line
Fix import simplejson as json
Moved service_type & auth_url in Connection

ZuluPro commented:
All your requests are now fixed at master of my fork.

I fixed the first line of License
Fix import statement, now with try/except,  import simplejson as json
I moved service_type & auth_url in Connection class

ZuluPro commented:
Do you want me to remake this patch from actual trunk ?

sebgoa commented:
Hi @ZuluPro sorry for the delay. Can you rebase on trunk and squash your commits. We will then look into merging it.

ZuluPro commented:
Ok, I will make !

Kami commented:
@ZuluPro Thanks.

Is it possible to get a test Cloudwatt account / credentials? I would like to test this driver before merging those changes into trunk.

ZuluPro commented:
@Kami, Sorry, my account is only for work purpose, Maybe you can ask them one account with notoriety of libcloud ?

Kami commented:
@ZuluPro I've made some changes so it works with the latest OpenStack auth refactoring (8aa3e27563970decd851ca84672562a41e53ef0a) and merged changes into trunk.

Next time please make sure you create a branch for your changes and don't directly commit to trunk - this will make it easier for both of us (this time, I needed to do some git foo and rebase to get it into mergable state).

Please also submit a new pull request which inclues documentation and usage examples for the driver (see `libcloud/compute/drivers`*.rst` for an examples) - driver without documentation (and tests) is not really useful.

Thanks.
